accounts/abi/bind/bindv2.go

```diff
@@ -0,0 +1,311 @@
+	errors map[string]*tmplError
+
+	callIdentifiers  map[string]bool
+	eventIdentifiers map[string]bool
+	errorIdentifiers map[string]bool
+}
+
+// bindMethod registers a method to be emitted in the bindings.
+// The name, inputs and outputs are normalized.  If any inputs are
+// struct-type their structs are registered to be emitted in the bindings.
+// Any methods that return more than one output have their result coalesced
+// into a struct.
+func (cb *contractBinder) bindMethod(original abi.Method) error {
+	normalized := original
+	normalizedName, err := cb.RegisterCallIdentifier(original.Name)
+	if err != nil {
+		return err
+	}
+
+	normalized.Name = normalizedName
+	normalized.Inputs = make([]abi.Argument, len(original.Inputs))
+	copy(normalized.Inputs, original.Inputs)
+	for j, input := range normalized.Inputs {
+		if input.Name == "" || isKeyWord(input.Name) {
+			normalized.Inputs[j].Name = fmt.Sprintf("arg%d", j)
+		}
+		if hasStruct(input.Type) {
+			cb.binder.BindStructType(input.Type)
+		}
+	}
+	normalized.Outputs = make([]abi.Argument, len(original.Outputs))
+	copy(normalized.Outputs, original.Outputs)
+	for j, output := range normalized.Outputs {
+		if output.Name != "" {
+			normalized.Outputs[j].Name = capitalise(output.Name)
+		}
+		if hasStruct(output.Type) {
+			cb.binder.BindStructType(output.Type)
+		}
+	}
+	isStructured := structured(original.Outputs)
+	// if the call returns multiple values, coallesce them into a struct
+	if len(normalized.Outputs) > 1 {
+		// Build up dictionary of existing arg names.
+		keys := make(map[string]struct{})
+		for _, o := range normalized.Outputs {
+			if o.Name != "" {
+				keys[strings.ToLower(o.Name)] = struct{}{}
+			}
+		}
+		// Assign names to anonymous fields.
+		for i, o := range normalized.Outputs {
+			if o.Name != "" {
+				continue
+			}
+			o.Name = capitalise(abi.ResolveNameConflict("arg", func(name string) bool { _, ok := keys[name]; return ok }))
+			normalized.Outputs[i] = o
+			keys[strings.ToLower(o.Name)] = struct{}{}
+		}
+		isStructured = true
+	}
+
+	cb.calls[original.Name] = &tmplMethod{Original: original, Normalized: normalized, Structured: isStructured}
+	return nil
+}
+
+// normalizeErrorOrEventFields normalizes errors/events for emitting through bindings:
+// Any anonymous fields are given generated names.
+func (cb *contractBinder) normalizeErrorOrEventFields(originalInputs abi.Arguments) abi.Arguments {
+	normalizedArguments := make([]abi.Argument, len(originalInputs))
+	copy(normalizedArguments, originalInputs)
+	used := make(map[string]bool)
+
+	for i, input := range normalizedArguments {
+		if input.Name == "" || isKeyWord(input.Name) {
+			normalizedArguments[i].Name = fmt.Sprintf("arg%d", i)
+		}
+		for index := 0; ; index++ {
+			if !used[capitalise(normalizedArguments[i].Name)] {
+				used[capitalise(normalizedArguments[i].Name)] = true
+				break
+			}
+			normalizedArguments[i].Name = fmt.Sprintf("%s%d", normalizedArguments[i].Name, index)
```

Comment:
this looks slightly wrong to me. We check for capitalized version of the name, and then iterate on changing the name. But we don't actually capitalize the name. So the used-check uses the capitalized version, but non-capitalized is set as name. 

It's a big confusing, and looks like a bug

Reply:
Shouldn't it be something like this? 
```golang
func normalizeArguments(args abi.Arguments) abi.Arguments {
	args := slices.Clone(args)
	used := make(map[string]bool)
	for i, input := range args {
		if input.Name == "" || isKeyWord(input.Name) {
			args[i].Name = fmt.Sprintf("arg%d", i)
		}
		name := capitalise(args[i].Name)
		for index := 0; ; index++ {
			if !used[name] {
				used[name] = true
				break
			}
			// try next
			name = capitalise(fmt.Sprintf("%s%d", args[i].Name, index))
		}
		args[i].Name = name
	}
	return args
}
```